### ISSUE TITLE
Culling mode with fixed zoom: Film Strip can jump back to start unexpectedly in relation to grouped images

### ISSUE THREAD

**Issue by pahl34** [darktable-org/darktable#6566]
**Update**:

Darktable version 3.4, Windows 10.

1. Load a collection of pictures shot in JPG and RAW.
2. Open culling mode with fixed zoom, with e.g. 2 images.
3. Expand grouped images.
4. Scroll a bit into the collection.
5. Select a JPG (click on it).
6. Collapse grouped images.

--> Filmstrip jumps back to the start and main view gets empty.

It might be argued whether this behavior is as designed, but it kind of breaks my workflow, so I wanted to mention it.
My expectation would be that after step 6 the filmstrip remains at the index of the RAW file (group head) that corresponds to the JPG selected in step 5.

-----------------------------------------------------------------------------------------------------------------------------------------
**Original Bug description for darktable version 3.2.1 with Windows 10.** The steps described in this original post lead to a darktable crash in version 3.4, but this crash is already documented in issue #7445. Keeping this part for the sake of completeness.

**Describe the bug**
In lighttable's culling mode with fixed zoom, if there are grouped images _and_ grouped images are collapsed (button in the top right of the main view part) and then the first ("leftmost") image of the currently shown ones is changed in rating, so that it does not match the current view filter any more, filmstrip "jumps back to the start" and there are no images shown any more until one is selected again in filmstrip. This is frustrating when reviewing a collection of images, because each time after this has happened it is required to first find the image to review next before it is possible to continue. When going through a bigger collection, this can take a lot of time and be quite annoying.

**_Please note_** again that this only happens if grouped images are collapsed - a fact that I only found out recently and after big parts of this error report had already been written. In the following, "image" therefore always refers to "group of images".

For the sake of clarification, the following is an attempt to rephrase the problem in a more formal way:

There is a collection of images currently active as output of the "collect images" module of lighttable, which I'll call **C**. Additionally, there is a view filter active, so that only those images out of **C** that match this filter are shown in the main view as well as in the filmstrip. This subset I'll call **C'**. **C'** is sorted in a certain way, thus it can be seen as a list of images with an index between 1 and **n**, with **n** being the total amount of images in **C'**.
The fixed zoom culling mode is set to show **k** images side by side, so for any selected image at index **i**, images **i** to **(i+k)** are shown.
If now the rating of image at index **i** is changed in a way so that this image is not any more part of **C'** (but only of **C**), then the whole view gets reset and no images are shown any more, as described above. The user needs to scroll through the filmstrip (i.e. through the sorted list that **C'** is) again in order to find the image that is now at **i** to be able to continue where she or he left off before.

**_Note:_**
- If the same is done for one of the other shown images (formally at index **j** with **i** < **j** <= **i+k**), then the problem does not occur - instead the image at **j** is replaced by the next in the list and the view does not get reset.
- The problem does not occur if the change of rating is done on the very first image in **C'** (i = 1).

**To Reproduce**
1. Open lighttable.
2. Load a collection of grouped images (e.g. images shot in JPG + RAW) using the "collect images" module.
3. Important: Make sure "collapse grouped images" is active (button in the top right of the main view).
4. Set a view filter, e.g. "All except rejected" (others can be used to reproduce the behavior if next steps are adapted accordingly)
5. Activate culling mode with fixed zoom and set the number of images to be shown next to each other to 2 (just for simplicitly, same behavior can also be reproduced for bigger number of pictures).
6. Scroll through the images so that the very first image in the list is not shown any more (the error does not seem to occur for that one), then move the mouse over the _left_ of the two shown images and press R. This will remove it from the current view, as it won't match the active filter any more.
**--> The error occurs, filmstrip jumps to the start and the main view gets emptied.**
7. In order to show that this only happens for the left one of the shown images, repeat the procedure, but in step 6 hover the mouse over the _right_ image and press 'R'. Now the error does not occur.

**Expected behavior**
I personally would expect the following behavior:
After removing an image from **C'** by changing its rating, the next image in the list should take its place, i.e. if image at index **i** to **(i+k)** are shown and then **i** gets removed, then the former images **(i+1) to (i+k+1)** should be shown (in fact, these are actually the new images **i to (i+k)** considering **C'**'s new content).

A special case is the one where images **(n-k) to n** are shown and image **(n-k)** gets removed from **C'**.  In this case, the former images **(n-k-1) to (n-1)** (the new images **(n-k) to n** considering **C'**'s new content) should be shown. Around the end of the list and depending on which value **k** has, there may be other corner cases that should be considered in a similar way.

**Screenshots**

**_Good case_**

![beforeRejectingRight](https://user-images.githubusercontent.com/60848211/96370635-cda9fe00-115e-11eb-987b-bd359ce7a9de.PNG)
Situation before pressing 'R' in order to reject the right image which will remove it from the current **C'**. The mouse cursor was on top of the right image (DSC_4000), but not captured on the screenshot unfortunately.

--------------------------------

![afterRejectingRight](https://user-images.githubusercontent.com/60848211/96370632-cd116780-115e-11eb-85a6-f9ea8d3bd4a1.PNG)
Situation immediately after pressing 'R', without any other interaction. This is the good case, behavior as expected. The next image (DSC_4003) has moved to the right slot that was freed by rejecting DSC_4000.

--------------------------------

**_Problematic case_**

![beforeRejectingLeft](https://user-images.githubusercontent.com/60848211/96370633-cda9fe00-115e-11eb-9f20-6af85de9c483.PNG)
Same situation as before, but this time the mouse cursor was on top of the left image (DSC_3998).

--------------------------------

![afterRejectingLeft](https://user-images.githubusercontent.com/60848211/96370631-cbe03a80-115e-11eb-8030-52e8e0583a08.PNG)
Situation immediately after pressing 'R', without any other interaction. This is the problematic case, filmstrip has "jumped back to the start", main view is empty.

--------------------------------

**Platform:**
 - Darktable Version: 3.2.1
 - OS: Windows 10

**Additional context**

The same matter has been discussed partly already as part of issue  #5735. However, it turned out, that:
1. This particular problem was out of context of that issue.
2. I was lacking some basic understanding of the culling mode and how it works, when I first posted there, which led to an imprecise and uncertain description.

I am now more confident that what I describe is indeed a reproducible behavior and worth its own issue entry. Also, I now found that there is a relation to the grouped images collapsed mode, as prominently mentioned above. This was probably the cause of my own confusion stated in the discussion of #5735 (I was suddenly not able to reproduce the problem any more, which was likely because I had expanded grouped images).

After last time I was (correctly) asked to read the manuals first, I'd like to state that there is of course still a risk that in the end I misunderstood some concept of the culling interface. However, after now reading the darktable 3.0 manual as well as the blog entries about 3.0 and 3.2.1 and spending more time with the culling mode, I still run into this problem regularly. Thus, even in case it _should_ indeed turn out to be my fault in the end, I would at least consider this issue a usability or documentation flaw.

**Comment by AlicVB** [comment]
@pahl34 : Thank you for this very detailed issue (maybe too detailed)
I can't reproduce, but as your are using 3.2.1 and there has been lot of fix in this area lately, I think this is already fixed in master.
So I'm closing, but don't hesitate to reopen or comment, if you manage to reproduce with current master.

Thanks for the report

**Comment by pahl34** [comment]
Just noticed that the same behavior can be reproduced the following way (still in darktable 3.2.1, windows 10).

1. Load a collection of pictures shot in JPG and RAW.
2. Open culling mode with fixed zoom, with e.g. 2 images.
3. Expand grouped images.
4. Scroll a bit into the collection.
5. Select a JPG (click on it).
6. Collapse grouped images.

--> Filmstrip jumps back to the start and main view gets empty.

Probably because the JPG that was last selected is not shown any more (with collapsed grouped images, the RAW images are group head and thus are shown instead of the whole group).

It might be argued whether this behavior is as designed, but it kind of breaks my workflow, so I wanted to mention it.
I think it would be nicer if the filmstrip remains at the index of the RAW file (group head) that corresponds to the JPG selected in step 5.

@AlicVB: Unfortunately I am a windows user who is not yet familiar with building the master. I'd like to learn it and maybe also switch to linux one day, but sadly I have very limited time currently (and spent a bit too much on my report above as you already hinted) so I could not yet dig deeper into these topics. Thus I am afraid I cannot offer to try to reproduce stuff with the current master atm. If you don't mind and have the chance to quickly try these new steps to reproduce on master, I would be very grateful, but of course I understand if you don't have time for it.

**Comment by pahl34** [comment]
I installed darktable 3.4 on my Windows 10 machine today and the problem is still reproducible for me the same way as described in my last post.

The steps described in my very long initial post now lead to a crash of darktable. I will open a new issue for that and attach the backtrace.

**Comment by johnny-bit** [comment]
@pahl34 - Please update your initial post then :) Issue is now reopened.

**Comment by pahl34** [comment]
@johnny-bit: Thanks for re-opening. I changed the title and the content of the initial post. However I decided for now to keep the original description as well for now, in case it could still be relevant.

**Comment by pahl34** [comment]
Tested with master 3.5.0~git273.1d772a434, which did not include your fix on Debain 10 - reproducible, as expected.
Updated to today's master (3.5.0~git294.9f83abb98) and cannot reproduce any more using the exact same steps. Even tried some corner cases, everything works nicely now.

@AlicVB: Thanks very much for the fix! This is extremely helpful for my culling workflow. :-)

**Comment by MeisterKaos** [comment]
Tested on Windows (3.5.0+324~gaa3f43133) to finally be able to get my rating workflow back. Works as expected in culling (THANKS!), but in file manager the behavour as described by @pahl34 in the comment on October 25 still persists. Could you double check, whether it is reproducible in file manager for you?

**Comment by pahl34** [comment]
At the moment I can only test master build 3.5.0~git294.9f83abb98 on Debian 10 - there I cannot reproduce any of the issues mentioned in this thread. No jumping back to the first image for me, neither with the steps described in the post from October 25, nor with the steps from my very long post with screenshots. Neither in culling, nor in file manager mode.

On windows 10 I currently only have 3.4 installed. Is the master build you mentioned available for windows somewhere or did you build it yourself?

**Comment by MeisterKaos** [comment]
@pahl34 I built the windows master myself according to the instructions provided here. It was a first time for me, previously I only used the stable versions. But I was hoping to completely get rid of this specific issue because it can really mess up my workflow^^ Unfortunately I don't know how to build a specific version (like the one you are using on debian) but only the most recent master.

**Comment by pahl34** [comment]
@MeisterKaos: Hm, in this case I can only assume that maybe you did not have AlicVB's commit [43db26e](https://github.com/darktable-org/darktable/pull/7569/commits/43db26edf4b0c4f35d279406b8697f4f966cf82a) in your local repo on build time..? Could that be possible? (However in that case it would not make sense that the issue was gone for you in culling mode...)

I tried to find commit 3f43133 (from your build label) but was not able to. Is it a local commit of yours?

Could you maybe also write down again the exact steps to reproduce the issue in file manager? Is it exactly what I wrote on October 25 or is there maybe a difference? I could then try once again to reproduce on my setup.

Otherwise, I'd suggest to wait for the 3.4.1 release and then test again on Windows, to be sure that there is no corner case left over.

**Comment by MeisterKaos** [comment]
Okay, thanks for the comment @pahl34 , I first have to learn how to do this. As I said I'm really new to using the git^^
'git show 43db26e' seems not to be working. 'git show 0857975' however successfully returns the commit. 

I was assuming that 'git clone git://github.com/darktable-org/darktable.git' automatically gets me into the most recent master and I did not change anything else before building. I now changed to the darktable-org:darktable-3.4.x branch and will try to build that one.
But since on AlicVB's commit that you linked it says 'TurboGit merged 1 commit into darktable-org:master from AlicVB:groups_bug1' I thought darktable-org:master would be the correct branch...

I'm sorry, this must be beginner mistakes. I hope I'll be learning quick :D

**Comment by MeisterKaos** [comment]
@pahl34 I updated the steps for reproduction in my last comment with an alternative way to reproduce the jumping (7-11). I also built a different version: 3.4.0+55~geb8d17139 which should be the current darktable-3.4.x
I still can reproduce both issues.

**Comment by pahl34** [comment]
@MeisterKaos: I cannot reproduce also with the updated steps...On Debian 10 with master build 3.5.0~git294.9f83abb98 (includes AlicVB's fix). What you describe actually sounds similar to the original problem, so maybe you indeed for some reason do not have the fix included..? 

One more question: I was a little bit confused by "Open file manager mode **with fixed zoom** 1 image."
The term "fixed zoom" seems to fit rather to culling mode (however, I also cannot reproduce in culling with 1 image). Just to be sure, you indeed meant the file manager, right?

Do you have a chance to test an 'official' build on Linux? I am also a Windows user, but I got it running using VirtualBox. Maybe this is also an option for you in case you do not have a linux machine available.

Also, maybe you could create a screencast/video of you reproducing the issue, it could help clarifying any minor differences that stilll may exist between what steps we take.

**Comment by MeisterKaos** [comment]
@pahl34 Sorry, the 'fixed zoom' was misleading. Just a copy paste from your steps. I am indeed talking about the file manager, displaying 1 image at a time (which basically is a fixed zoom, right?). I changed the description in my previous post.
Maybe I can find the time to do a screencast and/or install a vitualBox tomorrow.

**Comment by pahl34** [comment]
@MeisterKaos: I just updated to today's master (3.5.0~git480.4f8915dc3) and now I can reproduce the issue both ways you described in my Debian 10 VM.

Either I did something wrong with the steps earlier, or the bug was introduced very lately (I had used the master build from 01.01.2021 before).

Unfortunately I forgot to create a snapshot in my VirtualBox before, so I first have to downgrade to the earlier master build in order to double check.

As this issue (6566) is already closed and was related to culling instead of file manager, I would suggest to create a new one, maybe linking this one. In case I'll get to testing again the earlier master build I'll definitely report, as it might narrow down the commit that introduced this. However, I currently am afraid that it was just my mistake and the issue was present all the time.

**Comment by MeisterKaos** [comment]
All right, Thanks for the Update @pahl34 ! It seems you have not created a new issue yet? In that case I'll do it now.

**Comment by pahl34** [comment]
@AlicVB:
I installed darktable 3.4.1 lately (on windows 10). During usage (culling, fixed mode, 2 images, rejecting the first one of the two), the jumping back to the start of the current collection happened again several times. It does, however, only happen "sometimes" - I have not yet been able to find out what exactly I have to do to trigger the problem.

I double checked already whether your fix is actually included in 3.4.1 and it looked like it is (the commit is present on the darktable-3.4.x branch). So this looks like a corner case that still produces this issue.

Do you have any idea how this could still happen? Currently, I am a bit lost with finding the steps to reproduce, as there are so many possibilities. Maybe with you knowledge about the code, you can already rule out some of my theories:

Theories I had so far:
- Some relation to using the full screen preview ('W') before and scrolling there (it seems to have its own index)
- Some relation to switching from file manager to culling, then afterwards being able to reproduce the issue
- Some relation to having collected the images via the 'calendar selection' in the bottom left and not directly via the "collect images" module (I think this is unlikely, as it seems that the calendar selection is nothing else than creating a date-time-taken filter in the collect images module)

If this should have low prio for you, it is no problem - however I would really like to find out how to reproduce this, so that I can give you a proper report and you maybe get the chance to finally fix this issue in all its variations.
